### ISSUE TITLE
click.File ignores the errors argument when encoding is not specified

### ISSUE THREAD

**Issue by dlukes** [pallets/click#524]
See [click._compat.open_stream](https://github.com/mitsuhiko/click/blob/master/click/_compat.py#L429-L430):

``` python
if encoding is None:
    return open(filename, mode), True
```

Personally, I'm not sure why the case where the user doesn't explicitly specify an encoding is treated special, but I guess that's just because I'm unable to think through all of the compatibility ramifications :)

In any case, it would be nice to throw an error when a user tries to use the `errors` parameter on `click.File` without supplying `encoding`, because the current behavior is somewhat confusing:

``` python
#!/usr/bin/env python3

import click
import tempfile

with tempfile.NamedTemporaryFile(delete=False) as fh:
    fh.write("čáěšíčáž ".encode("utf-8") +
             bytearray.fromhex("ed") +
             " ěíéřáčéřž\n".encode("utf-8"))
    name = fh.name


@click.command()
@click.option("--input", type=click.File("r", errors="surrogateescape", encoding="utf-8"), default=name)
def test(input):
    click.echo(input.errors)
    for line in input:
        click.echo(line.encode("utf-8", errors="surrogateescape"))


if __name__ == "__main__":
    test()
```

When run this script outputs:

```
strict
Traceback (most recent call last):
  File "./test_click.py", line 22, in <module>
    test()
  File "/cnk/users/home/lukes/.linuxbrew/Cellar/python3/3.5.1/lib/python3.5/site-packages/click/core.py", line 716, in __call__
    return self.main(*args, **kwargs)
  File "/cnk/users/home/lukes/.linuxbrew/Cellar/python3/3.5.1/lib/python3.5/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/cnk/users/home/lukes/.linuxbrew/Cellar/python3/3.5.1/lib/python3.5/site-packages/click/core.py", line 889, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/cnk/users/home/lukes/.linuxbrew/Cellar/python3/3.5.1/lib/python3.5/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "./test_click.py", line 17, in test
    for line in input:
  File "/cnk/users/home/lukes/.linuxbrew/bin/../Cellar/python3/3.5.1/lib/python3.5/codecs.py", line 321, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xed in position 17: invalid continuation byte
```

Note `strict` on the first line. When `encoding="utf-8"` is specified for `click.File`, everything works as expected. I think some validation along the following lines in `click.types.File.__init__` would help guide users towards the right solution:

``` python
if encoding is None and errors != "strict":
    raise RuntimeError("Must specify ``encoding`` to set ``errors``.")
```

Python 3.5, Click 598ad3313bd02b43cdceb990cca8893242d43e63

**Comment by untitaker** [comment]
I suspect that special branch is a performance optimization. @mitsuhiko?

**Comment by davidism** [comment]
Closed by #1298.
